### DISCUSSION TITLE
Clean up unused imports

### DISCUSSION THREAD

**Discussion by nelsonsbrian** [biomejs/biome#3341] [Ideas]
I am requesting the feature of automatically removing unused imports from files. Not just empty/useless imports

When I'm coding, I have auto import that works great, now just need deletions of code blocks to be cleaned up. The efficiency lose is too great having to go back up to the top of the file to remove code.

I'm aware of the linter action https://biomejs.dev/linter/rules/no-unused-imports/

ESLint has this feature as a plugin with millions of downloads.  https://www.npmjs.com/package/eslint-plugin-unused-imports

Just being blunt, me using biome as a daily driver for work or personal use, the lack of feature here is a dealbreaker.

**Comment by ematipico** [comment]
If you're aware of the rule, which you can set up in your IDE, then I don't understand what you're requesting.

**Reply by nelsonsbrian** [reply]
Thanks for the feedback. I did work it out like you suggested. Just wasn't clear to me that you could and the old discussions didn't lead me to think that it was possible. Thanks!

**Comment by Conaclos** [comment]
You could apply safe fixes on save? This could automatically remove your unused imports,

**Reply by nelsonsbrian** [reply]
Thanks for the feedback. I did work it out like you suggested. Just wasn't clear to me that you could and the old discussions didn't lead me to think that it was possible. Thanks!

**Reply by Conaclos** [reply]
I think some docs is missing in the IDE integration page of our website. Under VSCode we could document the way of enabling auto-fixes. Also, we could point to it from `noUnusedImports` because it seems that many users are interested in removing unused imports on save.

**Reply by lukedukeus** [reply]
@Conaclos Where is the documentation on how to configure this?

**Reply by ematipico** [reply]
@lukedukeus https://biomejs.dev/linter/#configure-the-rule-fix

**Reply by ironsand-sou-eu** [reply]
@ematipico, can you help me? I configured the rule in biome.json as stated in those docs. My VSCode settings are like this:
<img width="358" height="102" alt="image" src="https://github.com/user-attachments/assets/4058b440-8920-4019-a464-ea8cb1f96d53" />

But it still don't remove unused imports on save.

**Comment by AlfianChabib** [comment]
<img width="344" height="119" alt="image" src="https://github.com/user-attachments/assets/57d557ff-1588-478d-ba74-28a522b4d60f" />

**Reply by ironsand-sou-eu** [reply]
Thank you SO much!

**Reply by hoangngoclam** [reply]
@AlfianChabib 
Thank you so much, it works for me
My setting is like this:
```json
  {
    "editor.codeActionsOnSave": {
      "source.fixAll.biome": "explicit",
    }
  }
 ```
